jupyterhub/tests/test_roles.py

```diff
@@ -1114,3 +952,78 @@ async def test_user_group_roles(app, create_temp_role):
     assert len(reply[0]['roles']) == 1
     assert reply[0]['name'] == 'jack'
     assert group_role.name not in reply[0]['roles']
+
+
+async def test_config_role_list():
+    roles_to_load = [
+        {
+            'name': 'elephant',
+            'description': 'pacing about',
+            'scopes': ['read:hub'],
+        },
+        {
+            'name': 'tiger',
+            'description': 'pouncing stuff',
+            'scopes': ['shutdown'],
+        },
+    ]
+    kwargs = {'load_roles': roles_to_load}
+    hub = MockHub(**kwargs)
+    hub.init_db()
+    hub.authenticator.admin_users = ['admin']
+    await hub.init_role_creation()
+    for role_conf in roles_to_load:
+        assert orm.Role.find(hub.db, name=role_conf['name'])
+    # Now reload and see if user is removed from role list
+    roles_to_load.pop(0)
+    hub.load_roles = roles_to_load
+    await hub.init_role_creation()
+    print(hub.db.query(orm.Role).all())
+    assert orm.Role.find(hub.db, name='tiger')
+    assert not orm.Role.find(hub.db, name='elephant')
+
+
+async def test_config_role_users():
+    role_name = 'painter'
+    user_name = 'benny'
+    user_names = ['agnetha', 'bjorn', 'anni-frid', user_name]
+    roles_to_load = [
+        {
+            'name': role_name,
+            'description': 'painting with colors',
+            'scopes': ['users', 'groups'],
+            'users': user_names,
+        },
+        {
+            'name': role_name,
+            'description': 'painting with colors',
+            'scopes': ['users', 'groups'],
+            'users': user_names,
+        },
+    ]
+    kwargs = {'load_roles': roles_to_load}
+    hub = MockHub(**kwargs)
+    hub.init_db()
+    hub.authenticator.admin_users = ['admin']
+    hub.authenticator.allowed_users = user_names
+    await hub.init_role_creation()
+    await hub.init_users()
+    await hub.init_role_assignment()
+    user = orm.User.find(hub.db, name=user_name)
+    role = orm.Role.find(hub.db, name=role_name)
+    assert role in user.roles
+    # Now reload and see if user is removed from role list
+    roles_to_load[0]['users'].remove(user_name)
```

Comment:
The 2nd definition of the same role will still have the `user_name` in `'users'` list but it gets ignored, correct?

Reply:
Yes indeed. I suppose that this is the way it should be, or should we implement duplicate roles to be prohibited?

Reply:
Good idea, I'll implement it to be prohibited then